bfsscp/bfsscp.go

```diff
@@ -0,0 +1,376 @@
+
+	// create a new iterator
+	return &iterator{
+		w:       b.client.Walk(b.withPrefix("/")),
+		pattern: b.withPrefix(pattern),
+	}, nil
+}
+
+// Head implements bfs.Bucket.
+func (b *bucket) Head(_ context.Context, name string) (*bfs.MetaInfo, error) {
+	fileInfo, err := b.client.Stat(b.withPrefix(name))
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	return &bfs.MetaInfo{
+		Name:    name,
+		Size:    fileInfo.Size(),
+		ModTime: fileInfo.ModTime(),
+	}, nil
+}
+
+// Open implements bfs.Bucket.
+func (b *bucket) Open(_ context.Context, name string) (bfs.Reader, error) {
+	f, err := b.client.Open(b.withPrefix(name))
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	fi, err := f.Stat()
+	if err != nil {
+		return nil, normError(err)
+	}
+
+	// Use file size to ensure limited read and avoid EOFs
+	return &readerCloser{r: f, size: fi.Size()}, nil
+}
+
+// Create implements bfs.Bucket.
+func (b *bucket) Create(ctx context.Context, name string, opts *bfs.WriteOptions) (bfs.Writer, error) {
+	f, err := ioutil.TempFile(b.config.TempDir, "bfs-scp")
+	if err != nil {
+		return nil, err
+	}
+
+	return &writer{
+		File:   f,
+		bucket: b,
+		ctx:    ctx,
+		name:   name,
+		opts:   opts,
+	}, nil
+}
+
+// Remove implements bfs.Bucket.
+func (b *bucket) Remove(_ context.Context, name string) error {
+	err := normError(b.client.Remove(b.withPrefix(name)))
+	if err != nil && err != bfs.ErrNotFound {
+		return err
+	}
+	return nil
+}
+
+// Close implements bfs.Bucket.
+func (b *bucket) Close() error {
+	return multierr.Combine(b.client.Close(), b.conn.Close())
+}
+
+// --------------------------------------------------------
+
+type writer struct {
+	*os.File
+
+	ctx    context.Context
+	bucket *bucket
+	name   string
+	opts   *bfs.WriteOptions
+
+	closeOnce sync.Once
+}
+
+func (w *writer) Discard() error {
+	err := os.ErrClosed
+	w.closeOnce.Do(func() {
+		// delete tempfile in the end
+		fname := w.Name()
+		defer os.Remove(fname)
+
+		// close tempfile
+		err = w.File.Close()
+	})
+	return err
+}
+
+func (w *writer) Commit() error {
+	err := os.ErrClosed
+	w.closeOnce.Do(func() {
+		// delete tempfile in the end
+		fname := w.Name()
+		defer os.Remove(fname)
+
+		// close tempfile, check context
+		if err = w.File.Close(); err != nil {
+			return
+		} else if err = w.ctx.Err(); err != nil {
+			return
+		}
+
+		// reopen for reading
+		var file *os.File
+		if file, err = os.Open(fname); err != nil {
+			return
+		}
+		defer file.Close()
+
+		fullName := w.bucket.withPrefix(w.name)
+		if err = w.bucket.client.MkdirAll(path.Dir(fullName)); err != nil {
+			return
+		}
+
+		var sf *sftp.File
+		sf, err = w.bucket.client.Create(fullName)
+		if err != nil {
+			return
+		}
+		_, err = io.Copy(sf, file)
+		return
+	})
+	return err
+}
+
+// --------------------------------------------------------
+
+type readerCloser struct {
+	r    *sftp.File
+	size int64
+}
+
+func (rc *readerCloser) Read(out []byte) (int, error) {
+	buf := make([]byte, rc.size)
```

Comment:
1. check ctx.Err() before anything else
2. this is not a very efficient way of doing this. you can just `return rc.r.Read(out)`, no?

Reply:
No, this entire structure is due to the underlying lib having an issue with reading. Our test fails and we get an unexpected EOF. (see `readerCloser` comment)

Reply:
Gone with `io.LimitReader` as per @mxmCherry 's suggestion